File: src/harness_core/dispatch/orchestration/task_decomposer.py
"""Task decomposer: breaks a TaskAnalysis into a WorkflowGraph."""

from __future__ import annotations

import uuid
from datetime import datetime, timezone

from ..task_analyzer import TaskAnalysis
from .agent_role_registry import AgentRoleRegistry
from .schemas import WorkflowEdge, WorkflowGraph, WorkflowNode


class TaskDecomposer:
    """Rule-based task decomposer. Produces a WorkflowGraph from a TaskAnalysis."""

    def __init__(self, role_registry: AgentRoleRegistry | None = None) -> None:
        self._registry = role_registry

    def decompose(self, analysis: TaskAnalysis, dispatch_id: str | None = None) -> WorkflowGraph:
        workflow_id = f"wf-{uuid.uuid4().hex[:12]}"
        now = datetime.now(timezone.utc).isoformat()

        nodes, edges = self._build_graph(workflow_id, analysis)

        return WorkflowGraph(
            workflow_id=workflow_id,
            dispatch_id=dispatch_id or analysis.analysis_id,
            nodes=tuple(nodes),
            edges=tuple(edges),
            status="decomposed",
            created_at=now,
            updated_at=now,
        )

    def _build_graph(
        self, workflow_id: str, analysis: TaskAnalysis
    ) -> tuple[list[WorkflowNode], list[WorkflowEdge]]:
        complexity = analysis.complexity_score

        if complexity < 0.3 and len(analysis.risk_flags) == 0:
            return self._simple_graph(workflow_id, analysis)

        if complexity >= 0.6 or len(analysis.risk_flags) >= 2:
            return self._complex_graph(workflow_id, analysis)

        return self._medium_graph(workflow_id, analysis)

    def _simple_graph(
        self, workflow_id: str, analysis: TaskAnalysis
    ) -> tuple[list[WorkflowNode], list[WorkflowEdge]]:
        node = self._make_node(workflow_id, analysis.task_domain, analysis)
        return [node], []

    def _medium_graph(
        self, workflow_id: str, analysis: TaskAnalysis
    ) -> tuple[list[WorkflowNode], list[WorkflowEdge]]:
        analyze = self._make_node(workflow_id, f"{analysis.task_domain}_analyze", analysis)
        execute = self._make_node(workflow_id, f"{analysis.task_domain}_execute", analysis, input_refs=(analyze.node_id,))
        edges = [WorkflowEdge(
            edge_id=f"edge-{uuid.uuid4().hex[:8]}",
            from_node_id=analyze.node_id,
            to_node_id=execute.node_id,
        )]
        return [analyze, execute], edges

    def _complex_graph(
        self, workflow_id: str, analysis: TaskAnalysis
    ) -> tuple[list[WorkflowNode], list[WorkflowEdge]]:
        analyze = self._make_node(workflow_id, f"{analysis.task_domain}_analyze", analysis)
        plan = self._make_node(workflow_id, f"{analysis.task_domain}_plan", analysis, input_refs=(analyze.node_id,))
        execute = self._make_node(workflow_id, f"{analysis.task_domain}_execute", analysis, input_refs=(plan.node_id,))
        review = self._make_node(workflow_id, f"{analysis.task_domain}_review", analysis, input_refs=(execute.node_id,))

        edges = [
            WorkflowEdge(edge_id=f"edge-{uuid.uuid4().hex[:8]}", from_node_id=analyze.node_id, to_node_id=plan.node_id),
            WorkflowEdge(edge_id=f"edge-{uuid.uuid4().hex[:8]}", from_node_id=plan.node_id, to_node_id=execute.node_id),
            WorkflowEdge(edge_id=f"edge-{uuid.uuid4().hex[:8]}", from_node_id=execute.node_id, to_node_id=review.node_id),
        ]
        return [analyze, plan, execute, review], edges

    def _make_node(
        self,
        workflow_id: str,
        task_type: str,
        analysis: TaskAnalysis,
        input_refs: tuple[str, ...] = (),
    ) -> WorkflowNode:
        node_id = f"node-{uuid.uuid4().hex[:8]}"

        agent_id = None
        if self._registry:
            agent_id = self._registry.assign_agent(workflow_id, node_id, task_type)

        budget = analysis.execution_budget_estimate / max(1, len(analysis.risk_flags) + 1)

        return WorkflowNode(
            node_id=node_id,
            workflow_id=workflow_id,
            task_type=task_type,
            assigned_agent_id=agent_id,
            input_refs=input_refs,
            budget=round(budget, 2),
            created_at=datetime.now(timezone.utc).isoformat(),
        )

File: engine/src/orchestration/task_decomposer.rs
use super::agent_role_registry::AgentRoleRegistry;
use super::schemas::{
    WorkflowEdge, WorkflowGraph, WorkflowNode, WORKFLOW_EDGE_SCHEMA_VERSION,
    WORKFLOW_NODE_SCHEMA_VERSION, WORKFLOW_SCHEMA_VERSION,
};
use crate::task_analyzer::TaskAnalysis;

pub struct TaskDecomposer {
    registry: Option<AgentRoleRegistry>,
}

#[allow(clippy::cloned_ref_to_slice_refs)]
impl TaskDecomposer {
    pub fn new(registry: Option<AgentRoleRegistry>) -> Self {
        Self { registry }
    }

    pub fn decompose(
        &mut self,
        analysis: &TaskAnalysis,
        dispatch_id: &str,
        runtime: &mut crate::runtime::FixtureRuntime,
    ) -> WorkflowGraph {
        let workflow_id = runtime.id("wf");
        let (nodes, edges) = self.build_graph(&workflow_id, analysis, runtime);

        WorkflowGraph {
            schema_version: WORKFLOW_SCHEMA_VERSION.to_string(),
            workflow_id,
            dispatch_id: dispatch_id.to_string(),
            nodes,
            edges,
            status: "decomposed".to_string(),
            created_at: runtime.now(),
            updated_at: runtime.now(),
            started_at: None,
            completed_at: None,
            result: None,
        }
    }

    fn build_graph(
        &mut self,
        workflow_id: &str,
        analysis: &TaskAnalysis,
        runtime: &mut crate::runtime::FixtureRuntime,
    ) -> (Vec<WorkflowNode>, Vec<WorkflowEdge>) {
        let complexity = analysis.complexity_score;

        if complexity < 0.3 && analysis.risk_flags.is_empty() {
            return self.simple_graph(workflow_id, analysis, runtime);
        }

        if complexity >= 0.6 || analysis.risk_flags.len() >= 2 {
            return self.complex_graph(workflow_id, analysis, runtime);
        }

        self.medium_graph(workflow_id, analysis, runtime)
    }

    fn simple_graph(
        &mut self,
        workflow_id: &str,
        analysis: &TaskAnalysis,
        runtime: &mut crate::runtime::FixtureRuntime,
    ) -> (Vec<WorkflowNode>, Vec<WorkflowEdge>) {
        let node = self.make_node(workflow_id, &analysis.task_domain, analysis, &[], runtime);
        (vec![node], vec![])
    }

    fn medium_graph(
        &mut self,
        workflow_id: &str,
        analysis: &TaskAnalysis,
        runtime: &mut crate::runtime::FixtureRuntime,
    ) -> (Vec<WorkflowNode>, Vec<WorkflowEdge>) {
        let analyze = self.make_node(
            workflow_id,
            &format!("{}_analyze", analysis.task_domain),
            analysis,
            &[],
            runtime,
        );
        let execute = self.make_node(
            workflow_id,
            &format!("{}_execute", analysis.task_domain),
            analysis,
            &[analyze.node_id.clone()],
            runtime,
        );
        let edge = WorkflowEdge {
            schema_version: WORKFLOW_EDGE_SCHEMA_VERSION.to_string(),
            edge_id: runtime.id("edge"),
            from_node_id: analyze.node_id.clone(),
            to_node_id: execute.node_id.clone(),
            edge_type: "dependency".to_string(),
        };
        (vec![analyze, execute], vec![edge])
    }

    fn complex_graph(
        &mut self,
        workflow_id: &str,
        analysis: &TaskAnalysis,
        runtime: &mut crate::runtime::FixtureRuntime,
    ) -> (Vec<WorkflowNode>, Vec<WorkflowEdge>) {
        let analyze = self.make_node(
            workflow_id,
            &format!("{}_analyze", analysis.task_domain),
            analysis,
            &[],
            runtime,
        );
        let plan = self.make_node(
            workflow_id,
            &format!("{}_plan", analysis.task_domain),
            analysis,
            &[analyze.node_id.clone()],
            runtime,
        );
        let execute = self.make_node(
            workflow_id,
            &format!("{}_execute", analysis.task_domain),
            analysis,
            &[plan.node_id.clone()],
            runtime,
        );
        let review = self.make_node(
            workflow_id,
            &format!("{}_review", analysis.task_domain),
            analysis,
            &[execute.node_id.clone()],
            runtime,
        );
        let edges = vec![
            WorkflowEdge {
                schema_version: WORKFLOW_EDGE_SCHEMA_VERSION.to_string(),
                edge_id: runtime.id("edge"),
                from_node_id: analyze.node_id.clone(),
                to_node_id: plan.node_id.clone(),
                edge_type: "dependency".to_string(),
            },
            WorkflowEdge {
                schema_version: WORKFLOW_EDGE_SCHEMA_VERSION.to_string(),
                edge_id: runtime.id("edge"),
                from_node_id: plan.node_id.clone(),
                to_node_id: execute.node_id.clone(),
                edge_type: "dependency".to_string(),
            },
            WorkflowEdge {
                schema_version: WORKFLOW_EDGE_SCHEMA_VERSION.to_string(),
                edge_id: runtime.id("edge"),
                from_node_id: execute.node_id.clone(),
                to_node_id: review.node_id.clone(),
                edge_type: "dependency".to_string(),
            },
        ];
        (vec![analyze, plan, execute, review], edges)
    }

    fn make_node(
        &mut self,
        workflow_id: &str,
        task_type: &str,
        analysis: &TaskAnalysis,
        input_refs: &[String],
        runtime: &mut crate::runtime::FixtureRuntime,
    ) -> WorkflowNode {
        let node_id = runtime.id("node");

        let agent_id = self
            .registry
            .as_mut()
            .and_then(|r| r.assign_agent(workflow_id, &node_id, task_type));

        let risk_divisor = (analysis.risk_flags.len() + 1) as f64;
        let budget = analysis.execution_budget_estimate as f64 / risk_divisor;

        WorkflowNode {
            schema_version: WORKFLOW_NODE_SCHEMA_VERSION.to_string(),
            node_id,
            workflow_id: workflow_id.to_string(),
            task_type: task_type.to_string(),
            assigned_agent_id: agent_id,
            status: "pending".to_string(),
            input_refs: input_refs.to_vec(),
            output_ref: None,
            budget: (budget * 100.0).round() / 100.0,
            cost_incurred: 0.0,
            error: None,
            created_at: runtime.now(),
            started_at: None,
            completed_at: None,
        }
    }
}
